Move to IRQBALANCE_BANNED_CPULIST

Tal-or opened the pull request:
<!--  Thanks for sending a pull request!

Please be aware that we're following the Kubernetes guidelines of contributing
to this project. This means that we have to use this mandatory template for all
of our pull requests.

Please also make sure you've read and understood our contributing guidelines
(https://github.com/cri-o/cri-o/blob/main/CONTRIBUTING.md) as well as ensuring
that all your commits are signed with `git commit -s`.

Here are some additional tips for you:

- If this is your first time, please read our contributor guidelines:
  https://git.k8s.io/community/contributors/guide#your-first-contribution and
  developer guide
  https://git.k8s.io/community/contributors/devel/development.md#development-guide
- Please label this pull request according to what type of issue you are
  addressing, especially if this is a release targeted pull request. For
  reference on required PR/issue labels, read here:
  https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
- If you want *faster* PR reviews, read how:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
- If the PR is unfinished, see how to mark it:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind dependency-change

<!--
Uncomment only one `/kind <>` line, hit enter to put that in a new line, and
remove leading whitespace from that line:
-->

<!--
/kind api-change
/kind bug
/kind ci
/kind cleanup
/kind dependency-change
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind feature
/kind flake
/kind other
-->

#### What this PR does / why we need it:
The IRQBALANCE_BANNED_CPUS variable is deprecated and will not be used anymore in the near future.

The newer IRQBALANCE_BANNED_CPULIST variable is already supported by irqbalanced (https://github.com/Irqbalance/irqbalance/blob/v1.9.0/cputree.c#L135) and should be used.

Implementing this requires new code for byte/hex formatted cpu mask to CPUSet formatting code.

#### Which issue(s) this PR fixes:
None

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->

<!--
Fixes #
or
None
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes see:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
None
```


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Refactor**
  * Internal rewrite to use CPU-list (cpuset) semantics across IRQ load‑balancing and CPU affinity flows; improved saving/restoring, error handling and logging; config now supports a CPULIST form for banned CPUs while public interfaces remain unchanged.
* **Tests**
  * Expanded unit and integration tests and helpers (including CPULIST conversion utilities and shell helpers) to validate CPULIST behavior, conversions, and save/restore/restart scenarios.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Tal-or committed:
utils: add helper functions

add helper functions for manipulating different CPU set
representaition.

The unit-tests have been adjusted for validating the
new helper functions.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>
Co-authored-by: Martin Sivak <msivak@redhat.com>

Tal-or committed:
utils: cleanup unused function

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or commented:
Supersede for: https://github.com/cri-o/cri-o/pull/8783

Tal-or commented:
/cc @MarSik

MarSik commented on internal/runtimehandlerhooks/utils_linux.go:
```diff
@@ -160,41 +149,47 @@ func isServiceEnabled(serviceName string) bool {
 	}
 	lines := strings.Split(string(input), "\n")
 	for _, line := range lines {
-		if strings.HasPrefix(line, irqBalanceBannedCpus+"=") {
```
I would read both to allow for seamless transition. You are commenting out the old variable above anyway.

Tal-or replied:
we definitely can read the legacy value, but then what shall we be doing with it?

MarSik replied:
Merge with the new one and then comment it out?

Tal-or replied:
Done. I added the logic and a comment that explains the new flow

Tal-or commented on internal/runtimehandlerhooks/utils_test.go:
```diff
@@ -145,4 +152,10 @@ const confTemplate = `# irqbalance is a daemon process that distributes interrup
+IRQBALANCE_BANNED_CPU_LIST=
 `
+
+func TestName(t *testing.T) {
+	c, _ := cpuset.Parse("0,3")
+	fmt.Printf("%q", toMask(32, c))
+}
```
Slipped inside. I should remove it.

Tal-or commented on internal/runtimehandlerhooks/high_performance_hooks_linux.go:
```diff
@@ -920,7 +921,7 @@ func RestoreIrqBalanceConfig(ctx context.Context, irqBalanceConfigFile, irqBanne
 	}
 
-	bannedCPUMasks, err := retrieveIrqBannedCPUMasks(irqBalanceConfigFile)
+	bannedCPUMasks, err := retrieveIrqBannedCPUList(irqBalanceConfigFile)
```
if we want to read it we should keep maintainig it

MarSik replied:
I think we should read it once in case tuned sets it (tuned only does that once, we disabled the dynamic management) and then we convert it to the newer one.

Tal-or replied:
Done.

Tal-or commented:
/retest-required

MarSik commented on internal/runtimehandlerhooks/utils_linux.go:
```diff
@@ -204,20 +208,31 @@ func updateIrqBalanceConfigFile(irqBalanceConfigFile, newIRQBalanceSetting strin
 	for _, line := range lines {
-		if strings.HasPrefix(line, irqBalanceBannedCpus+"=") {
-			return strings.Trim(strings.Split(line, "=")[1], "\""), nil
+		// try to read from the legacy file first
```
```suggestion
		// try to read from the legacy variable first
```

Just a nit.

MarSik commented on internal/runtimehandlerhooks/utils_linux.go:
```diff
@@ -204,20 +208,31 @@ func updateIrqBalanceConfigFile(irqBalanceConfigFile, newIRQBalanceSetting strin
+		if strings.HasPrefix(line, irqBalanceBannedCpusLegacy+"=") {
+			list := strings.Trim(strings.Split(line, "=")[1], "\"")
+
+			return mapHexCharToCPUSet(list)
```
What if both are present at the same time?

Tal-or replied:
It will preferred and go through the legacy flow first and only once. 

On the next iteration the legacy value will be commented out and the new value will be used.

MarSik commented on internal/runtimehandlerhooks/utils_linux.go:
```diff
@@ -228,3 +243,44 @@ func fileExists(filename string) bool {
+// calculateCPUSizeFromMask return the number of total cpus
+// given a CPU mask
+// this is useful for finding the complement of a mask.
+func calculateCPUSizeFromMask(mask string) int {
```
How will this behave in case the highest Id cpu is currently masked?

Tal-or replied:
AFAIR, the full mask will be always given even if the highest IDs are masked (are zeros)

Tal-or replied:
```
sh-5.1# cat /proc/irq/default_smp_affinity
ff,ffffffff,ffffffff,ffffffff

sh-5.1# echo ff0 > /proc/irq/default_smp_affinity
sh-5.1# cat /proc/irq/default_smp_affinity
00,00000000,00000000,00000ff0
```

MarSik replied:
Ok, I missed that indeed. If we assume the input is the full mask including zeros then this works fine.

MarSik commented on internal/runtimehandlerhooks/utils_linux.go:
```diff
@@ -228,3 +243,44 @@ func fileExists(filename string) bool {
+	noCommaMask := strings.ReplaceAll(mask, ",", "")
+	maskLen := len(noCommaMask)
+	// if mask is odd, round up
+	if maskLen%2 == 1 {
```
Why? Is even number of cores (no-HT case) a reliable assumption?

Tal-or replied:
fair enough

MarSik replied:
No need, just refreshed my memory about https://bugzilla.redhat.com/show_bug.cgi?id=2181546#c11 and rounding to the nearest byte boundary is fine. Just change the comment a bit to say something like:

"Make the hexadecimal mask full byte aligned in case the input ends with a half-byte."

Tal-or replied:
Discussing offline, according to https://bugzilla.redhat.com/show_bug.cgi?id=2181546#c11 it's fine

Tal-or committed:
irqbalance: Use the new IRQBALANCE_BANNED_CPULIST variable

The IRQBALANCE_BANNED_CPUS variable is deprecated and will not
be used anymore in the near future.

The newer IRQBALANCE_BANNED_CPULIST variable is already supported
by irqbalanced (https://github.com/Irqbalance/irqbalance/blob/v1.9.0/cputree.c#L135)
and should be used.

Implementing this requires new code for byte/hex formatted cpu mask
to CPUSet formatting code.

Co-authored-by: Martin Sivak <msivak@redhat.com>
Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or commented:
/retest

Tal-or commented:
/lifecycle/stale remove

Tal-or commented:
/retest

MarSik commented:
/retest

Tal-or commented:
/retest

MarSik commented:
/lgtm from me

Tal-or committed:
integration-tests: adjust irqbalance tests

Add the new `IRQBALANCE_BANNED_CPULIST` variable to the tests

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or commented:
> I don't understand the IRQ mask pieces but the code LGTM
> 
> though irq tests are now failing https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/cri-o_cri-o/8931/pull-ci-cri-o-cri-o-main-ci-cgroupv2-integration/1946141870413320192/artifacts/cgroupv2-integration/cri-o-gather/artifacts/testout.txt

@haircommander It's hard to understand the failure from the CI alone, is there a way to run the cgroupv2-integration tests locally?

Tal-or committed:
irqbalance: use - to mark empty banned cpu list

We can't use the empty string `` to mark empty banned list
because irqbalance daemon apply custom logic when it finds an empty string value.
see: https://github.com/Irqbalance/irqbalance/blob/8e8945e5092caf45605dfb1e66165e2eb9ec1f56/cputree.c#L145

This default behavior do not place IRQs on isolated or nohz_full cpus, which breaks the user intention.
To overcome the issue we're using this special value `-` to mark all cpus are participating in IRQ
balancing.

This is a temporary workaround until https://github.com/Irqbalance/irqbalance/pull/358 propagates d/s

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or commented:
/retest

Tal-or commented:
@haircommander I've fixed the cgroups test PTAL

Tal-or commented:
/lifecycle/stale remove

Tal-or committed:
unit-tests: adjust for CPUSet

Previous implementation used CPU mask.
Change to use CPUSet.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or commented:
/retest

MarSik commented:
/retest-required

MarSik reviewed (approved):
/lgtm

yanirq commented:
/remove-lifecycle stale

Tal-or commented:
/retest

Tal-or commented:
/retest

Tal-or commented:
/remove-lifecycle stale
waiting for: [RHEL-185722](https://redhat.atlassian.net/browse/RHEL-185722)

Tal-or commented:
/remove-lifecycle stale
waiting for: [RHEL-185722](https://redhat.atlassian.net/browse/RHEL-185722)
